### PR TITLE
Rect model

### LINKED ISSUE #25: Color Information for Paragraphs
By deeper paragraph analysis i want to get information about the paragraph background color, e.g. to check for the rendered code example.
At which part in textbox.py could i get color information from the PDFObj "behind" the scenes or is this to late?
Where could be a good point for it to get this information?

### PR DESCRIPTION
**Pull request by juiwenchen**
@kreuzberger I cherry picked your branch `upgrade` to this PR in order to make the change atomic. The credit for this rect feature is for you.

https://github.com/useblocks/libpdf/issues/25

### CODE REVIEW

**Comment by kreuzberger** [comment]
Hi! Thanks for integration!. If found a small issue in core.py

`LOG.info('Text rects crop: %s', 'no' if crop_rects_text else 'no')`
The first one should be a "yes"

But only affects logging, so very minor

File: libpdf/core.py [hunk 1]
```diff
@@ -106,20 +110,17 @@ def main(  # pylint: disable=too-many-arguments,too-many-locals  # no reasonable
...
+        LOG.info('Page range: [%s]', 'all' if not pages else ','.join(str(x) for x in pages))
+        LOG.info('Page crop: %s', 'not cropped' if not page_crop else ' '.join(str(x) for x in page_crop))
+        LOG.info('Smart page crop: %s', 'on' if smart_page_crop else 'off')
+        LOG.info('Extract annotations: %s', 'no' if no_annotations else 'yes')
+        LOG.info('Extract chapters: %s', 'no' if no_chapters else 'yes')
+        LOG.info('Extract paragraphs: %s', 'no' if no_paragraphs else 'yes')
+        LOG.info('Extract tables: %s', 'no' if no_tables else 'yes')
+        LOG.info('Extract figures: %s', 'no' if no_figures else 'yes')
+        LOG.info('Extract rects: %s', 'no' if no_rects else 'yes')
+        LOG.info('Text rects crop: %s', 'no' if crop_rects_text else 'no')
```
Comment (ubmarco) [inline libpdf/core.py]: as @kreuzberger already mentioned, is this a typo?
```suggestion
        LOG.info('Text rects crop: %s', 'yes' if crop_rects_text else 'no')
```

File: libpdf/extract.py [hunk 2]
```diff
@@ -639,6 +655,85 @@ def extract_figures(
...
+                    textboxes.append(hbox)
+
+                rect_name = f'page_{page.page_number}_rect.{idx_rect + 1}.png'
+
+                # create figures directory if not exist
+                Path(figure_dir).mkdir(parents=True, exist_ok=True)
+
+                rect_path = os.path.abspath(os.path.join(figure_dir, rect_name))
+
+                #figure = Figure(idx_figure + 1, image_path, fig_pos, links, textboxes, 'None')
```
Comment (ubmarco) [inline libpdf/extract.py]: commented code

**Review by ubmarco — CHANGES_REQUESTED** [review]
Good change, seems to work in a small test locally.
Please run `tox -e format`

**Comment by ubmarco** [comment]
Please rebase to check the new format-check tox env.

**Comment by juiwenchen** [comment]
> Hi! Thanks for integration!. If found a small issue in core.py
> 
> `LOG.info('Text rects crop: %s', 'no' if crop_rects_text else 'no')` The first one should be a "yes"
> 
> But only affects logging, so very minor

@kreuzberger Sorry, I didn't pay attention to the flag `crop_rects_text`. In our internal discussion, we would still like to consider the text in the coverage of rects as paragraphs or chapters. The reason is that chapters and paragraphs may also have the background color. I have adapted this concept in this RP, so rects are not excluded in chapter and paragraphs extraction. What do you think?

**Comment by kreuzberger** [comment]
> > Hi! Thanks for integration!. If found a small issue in core.py
> > `LOG.info('Text rects crop: %s', 'no' if crop_rects_text else 'no')` The first one should be a "yes"
> > But only affects logging, so very minor
> 
> @kreuzberger Sorry, I didn't pay attention to the flag `crop_rects_text`. In our internal discussion, we would still like to consider the text in the coverage of rects as paragraphs or chapters. The reason is that chapters and paragraphs may also have the background color. I have adapted this concept in this RP, so rects are not excluded in chapter and paragraphs extraction. What do you think?

@juiwenchen I think this is a good concept, rects should not be excluded in chapter and paragrapsh by default. This option was added cause the old figure handling behaved different (figure text was removed from paragraphs and chapters) and i wanted it therefore as an option. So this options could also be removed (crop_rects_text).

File: docs/contents/pdf_model.puml [hunk 3]
```diff
@@ -228,10 +233,12 @@
         HorizontalBox "+textbox  0..1" --* Cell
         HorizontalBox "+textboxes  0..*" --* Figure
         HorizontalBox "+textbox  0..1" --* Chapter
+        HorizontalBox "+textboxes  0..*" --* Rect
 
         Paragraph "+b_source  1" *-- "+links  *" Link
         Figure "+b_source  1" *-- "+links  *" Link
         Cell "+b_source  1" *-- "+links  *" Link
+        Rect "+b_source  1" *-- "+links  *" Link
```
Comment (ubmarco) [inline docs/contents/pdf_model.puml]: I think `Rect` cannot be a target of a link. It's just a graphical element.
If we do target search for links I think what people want is a paragraph, table or figure.
So if a `Rect` contains text, it is also a paragraph and the alorithm will find this in the search area.

**Review by ubmarco — CHANGES_REQUESTED** [review]

**Comment by juiwenchen** [comment]
> > > Hi! Thanks for integration!. If found a small issue in core.py
> > > `LOG.info('Text rects crop: %s', 'no' if crop_rects_text else 'no')` The first one should be a "yes"
> > > But only affects logging, so very minor
> > 
> > 
> > @kreuzberger Sorry, I didn't pay attention to the flag `crop_rects_text`. In our internal discussion, we would still like to consider the text in the coverage of rects as paragraphs or chapters. The reason is that chapters and paragraphs may also have the background color. I have adapted this concept in this RP, so rects are not excluded in chapter and paragraphs extraction. What do you think?
> 
> @juiwenchen I think this is a good concept, rects should not be excluded in chapter and paragrapsh by default. This option was added cause the old figure handling behaved different (figure text was removed from paragraphs and chapters) and i wanted it therefore as an option. So this options could also be removed (crop_rects_text).

@kreuzberger I finalized the PR. Apart from the above-mentioned change, I adjusted the rect model that only one textbox at maximum can be in the rect. In this case, only the text covered in the rect is extracted to a newly instantiated textbox as I don't know what is the best way to address the lt_textboxes which are overflowed the rect, so it is the simplest solution from our side. What do you think? 

The following is to summarize the changes in this PR based on your commit. If you are happy with this PR, do you mind running it against your test case and let us know if we should merge it.

- remove `crop_rects_text` as the text covered by the rect is considered as paragraphs or chapters
- extract the text covered by the rect to a newly instantiated textbox.

**Comment by kreuzberger** [comment]
Hi! I tried to test the branch, and it failed. 
It seems that the sphinx-simplepdf files itself will fail in the initial parse. This was fixed in the original pr

```
pdf = <pdfplumber.pdf.PDF object at 0x7fb9e9f3a350>

    def get_named_destination(pdf):  # pylint: disable=too-many-branches
        """
        Extract Name destination catalog.
    
        Extracts Name destination catalog (link target) from pdf.doc.catalog['Name'] to obtain
        the coordinates (x,y) and page for the corresponding destination's name.
    
        PDFPlumber does not provide explict 'Named Destinations of Document Catalog' like py2pdf, so it needs to be obtained
        by resolving the hierarchical indirect objects.
    
        The first step in this function is to check if the name destination exist in the PDF. If it does not, no extraction
        is executed.
    
        :param pdf: pdf object of pdfplumber.pdf.PDF
        :return: named destination dictionary mapping reference of destination by name object
        """
        LOG.info("Catalog extraction: name destination ...")
    
        # check if name tree exist in catalog and extract name tree
        name_tree = {}
        named_destination = {}
        pdf_catalog = pdf.doc.catalog
        if "Names" in pdf_catalog:
            # PDF 1.2
            if (
                isinstance(pdf_catalog["Names"], PDFObjRef)
                and "Dests" in pdf_catalog["Names"].resolve()
            ):
                name_tree = pdf_catalog["Names"].resolve()["Dests"].resolve()
            elif isinstance(pdf_catalog["Names"], dict) and "Dests" in pdf_catalog["Names"]:
>               name_tree = pdf_catalog["Names"]["Dests"].resolve()
E               AttributeError: 'dict' object has no attribute 'resolve
```

The orignal code in the PR was:
```
   if 'Names' in pdf_catalog:
        # PDF 1.2
        if isinstance(pdf_catalog['Names'], PDFObjRef) and 'Dests' in pdf_catalog['Names'].resolve():
            name_tree = pdf_catalog['Names'].resolve()['Dests'].resolve()
        elif isinstance(pdf_catalog['Names'], dict) and 'Dests' in pdf_catalog['Names']:
            name_tree = resolve1(pdf_catalog['Names']['Dests'])
            # name_tree = pdf_catalog['Names']['Dests'].resolve()
            # LOG.debug(f"{name_tree}")
```

@juiwenchen Suggestion: we add a test for it in this  repository on this branch?
[test_rects_extraction.pdf](https://github.com/useblocks/libpdf/files/14024285/test_rects_extraction.pdf)

by adding tests/test_rects.py and test this suggested pdf?

And how can i support this?

**Comment by juiwenchen** [comment]
> PDF 1.2

@kreuzberger  Interesting. I will have a quick fix from your original PR, and then we can merge this PR first. Afterwards, you can create a PR to add the test case for `test_rects_extraction.pdf`. What do you think

**Review by ubmarco — APPROVED** [review]

**Comment by kreuzberger** [comment]
i would suggest to add a testcase before merge to ensure the file could be opened
Could be simple. Here are the file contents:
```
"""Test rects extraction."""
from click.testing import CliRunner

import libpdf
from tests.conftest import (
    PDF_RECTS_EXTRACTION,
)

def test_rects_extraction():
    objects = libpdf.load(PDF_RECTS_EXTRACTION)
    assert objects.flattened.rects is not None
```
I would then add tests on a new PR. But also ok if i do it all later.

While running the tests i got severe problems executing them:
```
wand.exceptions.PolicyError: attempt to perform an operation not allowed by the security policy
```
I have to patch as superuser my etc/configs to get it running! See https://bugs.archlinux.org/task/60580
Was this due to changes in the code / libraries used? I didnt run into those problems before...

**Comment by kreuzberger** [comment]
@juiwenchen  ok, after latest update the test runs locally! You can merge, i add the tests later :smile: :+1:

**Pull request merged**